source/common/network/utility.cc

```diff
@@ -355,8 +355,19 @@ Address::InstanceConstSharedPtr Utility::getOriginalDst(Socket& sock) {
         sock.getSocketOption(SOL_IPV6, IP6T_SO_ORIGINAL_DST, &orig_addr, &addr_len).return_value_;
   }
 
+#ifdef IP_TRANSPARENT
+  socklen_t flag_len = sizeof(int);
+  int is_transparent;
+  if (status != 0) {
+    status = sock.getSocketOption(SOL_IP, IP_TRANSPARENT, &is_transparent, &flag_len).return_value_;
+    if (status != 0 || !is_transparent) {
+      return nullptr;
+    }
+    safeMemcpy(reinterpret_cast<sockaddr*>(&orig_addr), sock.ioHandle().localAddress()->sockAddr());
```

Comment:
It is not immediately obvious to me that `orig_addr` would be safe to use here. Is it possible that status on line 352 is error on IP_TRANSPARENT? Is there some other reason besides IP_TRANSPARENT that could make the address invalid?

Reply:
> Is it possible that status on line 352 is error on IP_TRANSPARENT? 

Yes.  SO_ORIGINAL_DST is implemented in nf_conntrack, so status on line 352 would be -1 when there is no nf_conntrack module.

> Is there some other reason besides IP_TRANSPARENT that could make the address invalid?

I think original_dst filter implies listener socket is not `bind_to_port`,  in another word, traffic is redirected to listener socket.

Normally we use iptables NAT targets (including SNAT/DNAT/REDIRECT/MASQUERADE) to do that, which relies on nf_conntrack. So there is always nf_conntrack module in this "traffic-redirected-by-NAT" scenario, and getting SO_ORIGINAL_DST option works fine.

[tproxy](https://www.kernel.org/doc/html/latest/networking/tproxy.html#redirecting-traffic) is another way to do traffic redirect, which requires listener socket is IP_TRANSPARENT and could work without nf_conntrack. Then `getSocketOption` on line 352 returns -1 and func `Utility::getOriginalDst` returns nullptr. 

In this  "traffic-redirected-by-only-tproxy" scenario, the original_dst is actually the accepted socket's local address. This patch is trying to screen this branch out as a valid original_dst address.

This is my consideration, please let me know if it's wrong or confusing.

Reply:
It seems like you can return sock.ioHandle().localAddress() directly, needn't a memcopy to orig_addr

Reply:
yes.. got it